---
title: 'CRISPR sgRNA Library composition: Yusa v3 (Knockout)'
author: "Alex Kalinka"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: hide
    depth: 3
    highlight: tango
    number_sections: true
    theme: spacelab
    toc: true
    toc_float: true
---


```{r,echo=F}
options(warn=-1)

suppressMessages(library(knitr))
suppressMessages(library(tidyr))
suppressMessages(library(magrittr))
suppressMessages(library(readr))
suppressMessages(library(ggplot2))
suppressMessages(library(dplyr))

select <- dplyr::select
filter <- dplyr::filter

base_path.data <- file.path("~","data","az_cruk")
base_path.ht29 <- file.path(base_path.data,"ht29")

# 'cleanr.tsv' file from Miika.
yusa.v3 <- suppressMessages(suppressWarnings(read_table2(file.path(base_path.data,"cleanr.tsv")))) %>%
  rowwise() %>%
  mutate(GC_percent = 100*(sum(unlist(strsplit(seq,"")) == "G") + sum(unlist(strsplit(seq,"")) == "C"))/19) %>%
  ungroup()

# Cached since expensive to re-create.
yusa.annot <- readRDS(file.path(base_path.data,"yusa.v3.annotated.rds"))

# Sanger pan-cancer core fitness (essential genes).
pcf <- read.csv(file.path(base_path.data,"pan_cancer_core_fitness_genes--Sanger.csv"),
                  stringsAsFactors = F)
# Ensembl tx IDs.
ens_tx_ids <- readRDS(file=file.path(base_path.data,"ensembl_tx_ids.rds"))

```

# Annotations

* Variance in counts between guides within genes vs number of transcripts impacted.

## Setup

```{r,eval=F}
suppressMessages(library(GenomicRanges))
suppressMessages(library(annotatr))

# Annotations of interest.
anns <- c("hg38_genes_promoters","hg38_genes_cds","hg38_genes_exons",
          "hg38_genes_5UTRs","hg38_genes_3UTRs",
          "hg38_genes_firstexons",
          "hg38_genes_intronexonboundaries",
          "hg38_genes_exonintronboundaries")
annotations <- build_annotations(genome = "hg38", annotations = anns)

# Process sgRNA coordinates into GRanges object.
yusa.granges <- yusa.v3 %>%
  filter(!grepl("Control",CODE)) %>%
  select(CHRM,STARTpos,ENDpos,STRAND,CODE,GENES,seq) %>%
  mutate(CHRM = paste("chr",CHRM,sep=""),
         STARTpos = as.integer(STARTpos), ENDpos = as.integer(ENDpos)) %>%
  makeGRangesFromDataFrame(keep.extra.columns = T,
                           seqnames.field = "CHRM",
                           start.field = "STARTpos",
                           end.field = "ENDpos",
                           strand.field = "STRAND")

# Annotate sgRNAs.
yusa.annot <- data.frame(annotate_regions(regions = yusa.granges, annotations = annotations, 
                               ignore.strand = T, quiet = F)) %>%
  filter(!is.na(annot.symbol)) %>%
  mutate(feature = gsub("^(\\S+?):.*$","\\1",annot.id)) %>%
  group_by(CODE) %>%
  summarise(gene = GENES[1], 
            genes.hit = paste(unique(annot.symbol),collapse=";"),
            chrom = seqnames[1], start = start[1], end = end[1],
            seq = seq[1], 
            num_transcripts_hit = length(unique(annot.tx_id[which(annot.symbol==gene)])),
            num_genes = length(unique(annot.symbol)),
            features = paste(sort(unique(feature)),collapse=";"),
            CDS = grepl("CDS",features),
            promoter = grepl("promoter",features),
            transcript_ids_hit = paste(unique(gsub("^(.*?)\\..*$","\\1",
                                                   annot.tx_id[which(annot.symbol==gene)])),
                                       collapse=";")) %>%
  ungroup()

```

```{r,eval=F}
# Get the total number of transcripts annotated to each gene.
suppressMessages(library(biomaRt))
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")

ens_tx_ids <- getBM(filters = c("external_gene_name"),
                 values = yusa.annot$gene,
                 attributes = c("external_gene_name","ensembl_gene_id","ensembl_transcript_id",
                                "chromosome_name"),
                 mart = ensembl) %>%
  filter(!grepl("CHR",chromosome_name)) %>%
  group_by(ensembl_gene_id) %>%
  summarise(gene = external_gene_name[1],
            transcript_ids = paste(ensembl_transcript_id,collapse=";"),
            total_num_transcripts = n()) %>%
  ungroup()

# Finalise sgRNA annotations.
yusa.annot %<>%
  left_join(ens_tx_ids, by = "gene") %>%
  mutate(num_transcripts_not_hit = ifelse(num_transcripts_hit >= total_num_transcripts,
                                          0, total_num_transcripts - num_transcripts_hit),
         percent_transcripts_not_hit = 100*num_transcripts_not_hit/total_num_transcripts) %>%
  # Summarise number of tx not hit per gene.
  group_by(gene) %>%
  mutate(total_tx_not_hit.per_gene = sum(num_transcripts_not_hit),
         median_tx_not_hit.per_gene = median(num_transcripts_not_hit)) %>%
  ungroup()

```

## Post-selection median logFC: HT29 lethality screen

```{r}
# FGC_AZ_001_1 and FGC_AZ_001_2: Controls
ht29_ps_lfc <- read.delim(file.path(base_path.ht29,"combined_counts.txt")) %>%
  select(-gene) %>%
  rowwise() %>%
  mutate(Control_Plasmid.logFC.median = log2(median(c(FGC_AZ_001_1,FGC_AZ_001_2))) - 
                                               log2(median(c(FGC_AZ_001_3,FGC_AZ_001_4)))) %>%
  ungroup() %>%
  inner_join(yusa.annot, by = c("sgRNA" = "CODE")) %>%
  inner_join(pcf, by = "gene")

# Correlation of tx_not_hit & median log FC within genes.
ht29_lfs_miss_tx.cor <- ht29_ps_lfc %>%
  group_by(gene,ensembl_gene_id) %>%
  filter(sd(num_transcripts_not_hit)!=0 & Control_Plasmid.logFC.median!=Inf & Control_Plasmid.logFC.median!=(-Inf) & !is.na(Control_Plasmid.logFC.median) &
           !is.na(num_transcripts_not_hit)) %>%
  filter(n() >= 4) %>%
  summarise(num_transcripts_not_hit.median = median(num_transcripts_not_hit,na.rm=T),
            pct_transcripts_not_hit.median = median(percent_transcripts_not_hit,na.rm=T),
            median.logFC.gene = median(Control_Plasmid.logFC.median, na.rm=T),
            sd.logFC.gene = sd(Control_Plasmid.logFC.median, na.rm=T),
            num_transcripts_not_hit.sd = sd(num_transcripts_not_hit,na.rm=T),
            cor.lfc_num_tx_not_hit.spearman = cor(Control_Plasmid.logFC.median, num_transcripts_not_hit, method = "spearman"),
            cor.lfc_num_tx_not_hit.spearman.pvalue = cor.test(Control_Plasmid.logFC.median, num_transcripts_not_hit, method = "spearman")$p.value,
            random.cor.lfc_num_tx_not_hit.spearman = cor(
              sample(Control_Plasmid.logFC.median,n(),replace=F),
              sample(num_transcripts_not_hit,n(),replace=F), method = "spearman"),
            PanCancer_Core_Fitness = PanCancer_Core_Fitness[1],
            bagel_essential = bagel_essential[1], Ribosomal_Proteins = Ribosomal_Proteins[1]) %>%
  ungroup()

# Examples.
plot_lfc <- function(data, tgene){
  pl <- ggplot(data %>% filter(gene==tgene),
         aes(percent_transcripts_not_hit,Control_Plasmid.logFC.median)) +
    geom_point() +
    ggtitle(tgene)
  print(pl)
}


genes <- c("ANAPC11", "DDX46", "CENPK", "AURKA", "EIF2B4", "PHF8", "SNF8", "POLR1D", "MAK16", "BRD4")
for(tgene in genes){
  plot_lfc(ht29_ps_lfc,tgene)
}

```


## Post-selection Bayes Factors

```{r}
# Bayes Factors for HT29 lethality screen.
ht29_BF <- read.delim(file.path(base_path.ht29,"crisprn","plasmid_test","bagel",
                                "Control_vs_Plasmid.bf"), header=T, stringsAsFactors = F) %>%
  inner_join(yusa.annot, by = c("GENE" = "gene")) %>%
  inner_join(pcf, by = c("GENE" = "gene"))

```


# GC content vs counts

```{r}
plasmid.1 <- read.table(file.path(base_path.ht29,"crisprn","plasmid_test","counts","Plasmid_AZ_1.s_1.counts"),
                        stringsAsFactors = F) %>%
  left_join(yusa.v3, by = c("V1" = "CODE")) %>%
  group_by(GC_percent) %>%
  mutate(median_count = as.integer(round(median(V3,na.rm=T)))) %>%
  ungroup()

ggplot(plasmid.1, aes(GC_percent, V3)) +
  geom_count(aes(color = ..n..)) +
  stat_summary(aes(y = V3), fun.y = median, color = "red", geom="point") +
  geom_smooth() +
  ylim(0,2000) +
  ylab("sgRNA count") +
  ggtitle("HT29: Plasmid Count rep 1 vs GC %")

```

